eng/native/configurecompiler.cmake

```diff
@@ -532,20 +532,40 @@ if (CLR_CMAKE_HOST_UNIX)
     endif()
   endif()
 
-  if(CLR_CMAKE_HOST_OSX OR CLR_CMAKE_HOST_MACCATALYST)
-    # We cannot enable "stack-protector-strong" on OS X due to a bug in clang compiler (current version 7.0.2)
-    add_compile_options(-fstack-protector)
-    if(CLR_CMAKE_HOST_UNIX_ARM64)
-      # For OSX-Arm64, LSE instructions are enabled by default
-      add_definitions(-DLSE_INSTRUCTIONS_ENABLED_BY_DEFAULT)
-      add_compile_options(-mcpu=apple-m1)
-    endif(CLR_CMAKE_HOST_UNIX_ARM64)
-  elseif(NOT CLR_CMAKE_HOST_BROWSER AND NOT CLR_CMAKE_HOST_WASI)
+  if((CLR_CMAKE_HOST_OSX OR CLR_CMAKE_HOST_MACCATALYST) AND CLR_CMAKE_HOST_UNIX_ARM64)
+    # For osx-arm64, LSE instructions are enabled by default
+    add_definitions(-DLSE_INSTRUCTIONS_ENABLED_BY_DEFAULT)
+    add_compile_options(-mcpu=apple-m1)
+  endif()
+
+  # hardening options
+  if(NOT CLR_CMAKE_HOST_BROWSER AND NOT CLR_CMAKE_HOST_WASI)
     check_c_compiler_flag(-fstack-protector-strong COMPILER_SUPPORTS_F_STACK_PROTECTOR_STRONG)
     if (COMPILER_SUPPORTS_F_STACK_PROTECTOR_STRONG)
       add_compile_options(-fstack-protector-strong)
     endif()
-  endif(CLR_CMAKE_HOST_OSX OR CLR_CMAKE_HOST_MACCATALYST)
+
+    check_c_compiler_flag(-fstack-clash-protection COMPILER_SUPPORTS_F_STACK_CLASH_PROTECTION)
+    # explicit check for gcc is required because while clang does not implement stack-protection,
+    # it does not return 'unknown argument' either, instead:
+    #   clang: warning: argument unused during compilation: '-fstack-clash-protection' [-Wunused-command-line-argument]
+    # see:
+    #   https://github.com/llvm/llvm-project/issues/40148
+    #   https://gitlab.kitware.com/cmake/cmake/-/issues/25390
+    if (COMPILER_SUPPORTS_F_STACK_CLASH_PROTECTION AND CMAKE_C_COMPILER_ID STREQUAL "GNU")
+      add_compile_options(-fstack-clash-protection)
+    endif()
+
+    check_c_compiler_flag(-fcf-protection COMPILER_SUPPORTS_F_CONTROL_FLOW_PROTECTION)
+    if (COMPILER_SUPPORTS_F_CONTROL_FLOW_PROTECTION)
+      add_compile_options(-fcf-protection)
+    endif()
+
+    # Enable maximum fortification
+    string(REGEX REPLACE "-D_FORTIFY_SOURCE=[0-9] " "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
```

Comment:
Why do we need to throw away what was passed in? It does not look right to ignore conflicting settings silently.

Reply:
It is not new, we moved it from configureplatform.cmake.

Reply:
Instead of throwing away the options, do we want to use the existing option if it's already specified by the platform build scripts?

I've noticed that some community distro builds have scripts to manipulate these flags because their default settings conflict with ours.

Reply:
Yes, I think it will be better to honor the distro choice here. We can scan the `CMAKE_C{XX}_FLAGS` for `_FORTIFY_SOURCE` and set the default value only when it is not set by caller.